packages/circuits/circuits/bigComparators.circom

```diff
@@ -6,13 +6,13 @@ include "./modulo.circom";
 //~~ support comparisons of numbers up to the field size
 
 template UpperLessThan(n) {
-    assert(n < 254);
+    assert(n < 253);
     signal input in[2];
     signal output out;
 
     component bits[2];
     for (var x = 0; x < 2; x++) {
-        bits[x] = Num2Bits_strict();
+        bits[x] = Num2Bits(253);
```

Comment:
If the circuit can handle the bits, the following code doesn't have to check like
```ts
random() % BigInt(2) ** BigInt(253)
```

Reply:
Can you explain more about how it's shifted? Like how are you converting to bits?

Reply:
or 
```circom
signal shifted_field <-- 21888242871839275222246405745257275088548364400416034343698204186575808495616 >> (254-n);
...
field_lt[x] = LessEqThan(n);
```
Use `SNARK_FIELD-1` and `LessEqThan` if using `SNARK_FIELD` is dangerous

Reply:
I also found that for some certain large `in[x]` (e.g. `(SNARK_FIELD << 3) + 1`)
even `Num2Bits_strict()` cannot detect the overflow 👀

Reply:
This is interesting, maybe we could do like

```js
    component bits[2];
    signal shifted_bits[2];
    signal lower_bits[2];
    for (var x = 0; x < 2; x++) {
        bits[x] = Num2Bits(n);
        shifted_bits[x] <-- in[x] >> (254-n);
        lower_bits[x] <-- in[x] & (254-n);
        in[x] === shifted_bits[x] * (2**(254-n)) + lower_bits[x];
        bits[x].in <== shifted_bits[x];
    }

```

> I also found that for some certain large in[x] (e.g. (SNARK_FIELD << 3) + 1)
even Num2Bits_strict() cannot detect the overflow 👀

If you supply a value larger than `SNARK_FIELD` as an input circom should automatically apply `% SNARK_FIELD`. You can't overflow by providing a bad input, only by manually modifying the witness.

Reply:
Actually I think none of these are safe.

In this one the `shifted_bits[x]` signal is unconstrained, so it can be chosen as anything.

```
    component bits[2];
    signal shifted_bits[2];
    for (var x = 0; x < 2; x++) {
        bits[x] = Num2Bits(n);
        shifted_bits[x] <-- in[x] >> (254-n);
        bits[x].in <== shifted_bits[x];
    }
```

In this one the `shifted_bits[x]` is constrained to be less than `SNARK_FIELD >> (254-n)`, but it can be any value in that range. Also `shifted_field` could be assigned as `<==` because the rhs is constant.

```
    component bits[2];
    signal shifted_field <-- 21888242871839275222246405745257275088548364400416034343698204186575808495617 >> (254-n);
    signal shifted_bits[2];
    component field_lt[2];
    for (var x = 0; x < 2; x++) {
        bits[x] = Num2Bits(n);
        field_lt[x] = LessThan(n);
        shifted_bits[x] <-- in[x] >> (254-n);
        field_lt[x].in[0] <== shifted_bits[x];
        field_lt[x].in[1] <== shifted_field;
        field_lt[x].out === 1;
        bits[x].in <== shifted_bits[x];
    }
```

In this one there should be many pairs of `shifted_bits[x]`/`lower_bits[x]` that fulfill `in[x] === shifted_bits[x] * (2**(254-n)) + lower_bits[x]`. e.g. if `in[0] == 1` we could take large values fo `shifted_bits[x]` and `lower_bits[x]` and overflow to equal `in[0]`. This might be safe based on schwartz-zippel though - the odds of finding a collision might be very low. It's probably worth asking the auditors.

```
    component bits[2];
    signal shifted_bits[2];
    signal lower_bits[2];
    for (var x = 0; x < 2; x++) {
        bits[x] = Num2Bits(n);
        shifted_bits[x] <-- in[x] >> (254-n);
        lower_bits[x] <-- in[x] & (254-n);
        in[x] === shifted_bits[x] * (2**(254-n)) + lower_bits[x];
        bits[x].in <== shifted_bits[x];
    }

```

Reply:
> In this one the `shifted_bits[x]` signal is unconstrained, so it can be chosen as anything.

I think it is constrained by
```circom
        bits[x] = Num2Bits(n);
        bits[x].in <== shifted_bits[x];
```
so it should be `n` bits
like we constrain epoch, nonce,...

> In this one the `shifted_bits[x]` is constrained to be less than `SNARK_FIELD >> (254-n)`, but it can be any value in that range.

yeah it is designed to let the `shifted_bits[x]` to be less then `SNARK_FIELD >> (254-n)`
so even if it is within `n` bits, it can detect the overflow

> Also `shifted_field` could be assigned as `<==` because the rhs is constant.

yeah I agree with this

> we could take large values for `shifted_bits[x]` and `lower_bits[x]`

you mean if `in[x]` is overflow we cannot detect by 
```circom
in[x] === shifted_bits[x] * (2**(254-n)) + lower_bits[x];
```
?

and actually there is a small bug
```circom
lower_bits[x] <-- in[x] & (254-n);
```
should be
```circom
lower_bits[x] <-- in[x] & (2**(254-n) - 1);
```

Reply:
```js
        bits[x] = Num2Bits(n);
        shifted_bits[x] <-- in[x] >> (254-n); // any value may be chosen here
        bits[x].in <== shifted_bits[x];
```

This only constrains the values to be less than `2**n`. The user can still choose any value they want (between `[0,2**n)` for `shifted_bits[x]` because the assignment above isn't constrained.

Reply:
could you show me some example?
How can the `shifted_bits[x]` be assigned (and different from `in[x] >> (254-n)`)?

Reply:
Say that we have this circuit

```js
pragma circom 2.0.0;

template Example() {
  signal input in;
  signal output out;

  signal shifted_value <-- in >> 234;
  shifted_value === 9;

  out <== in;
}

component main = Example();
```

This circuit should only allow 1 value for `in`, which should be `9 << 234`. So I build the circuit and then create a witness. The witness looks like this:

```
7774 6e73 0200 0000 0200 0000 0100 0000
2800 0000 0000 0000 2000 0000 0100 00f0
93f5 e143 9170 b979 48e8 3328 5d58 8181
b645 50b8 29a0 31e1 724e 6430 0200 0000
0200 0000 4000 0000 0000 0000 0100 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0024 0000
```

You can see at the end the `out` value is `240000000000000000000000000000000000000000000000000000000000`, which is the hex value of `9 << 234`.

Now I can change the `out` value to anything I want because there are no constraints on the value `in`, only on `shifted_value`, which is not constrained (assigned using `<--`).

So I change the witness to 

```
7774 6e73 0200 0000 0200 0000 0100 0000
2800 0000 0000 0000 2000 0000 0100 00f0
93f5 e143 9170 b979 48e8 3328 5d58 8181
b645 50b8 29a0 31e1 724e 6430 0200 0000
0200 0000 4000 0000 0000 0000 0100 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 001c 0000
```

Which gives me `out == 7 << 234`. snarkjs will generate a valid proof using this witness, and the public signal `out` will be `7 << 234`.

Reply:
Also, public inputs don't have constraints by default. For this circuit

```js
pragma circom 2.0.0;

template Example() {
  signal input in;
  signal output out;
  signal input data;

  signal shifted_value <-- in >> 234;
  shifted_value === 9;

  out <== in;
}

component main { public [ data ] } = Example();
```

With this input

```json
{
  "in": "248462868484460296347651211041972916288239922998133915985850872056774656",
  "data":"222"
}
```

we get witness

```
7774 6e73 0200 0000 0200 0000 0100 0000
2800 0000 0000 0000 2000 0000 0100 00f0
93f5 e143 9170 b979 48e8 3328 5d58 8181
b645 50b8 29a0 31e1 724e 6430 0300 0000
0200 0000 6000 0000 0000 0000 0100 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0024 0000 de00 0000
0000 0000 0000 0000 0000 0000 0000 0000
0000 0000 0000 0000 0000 0000 
```

I can freely change the data input (`de00`) to whatever I choose and make a proof. So i guess we still need the dummy squares

Reply:
oh I didn't know that we can change the witness to generate a valid proof
yeah we can try dummy squares